TestMathBits: Call functions with original sample input values

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69026
* #68963
* #68950
* #68949
* #68948
* __->__ #68947

`_test_math_view` currently calls the operator with different values
than those specified in the `SampleInput`. This is undesirable as it
could break mathematical properties required by the operator. Instead,
this calls `math_op_view(math_op_physical(sample.input))` to get a
view that represents the same value as the original input.

`test_neg_view` already did this by returning `torch._neg_view(-x)`
from `math_op_view` but this moves the handling into `_test_math_view`
to make it apply to all view op tests.

Differential Revision: [D33064327](https://our.internmc.facebook.com/intern/diff/D33064327)

peterbell10 committed:
TestMathBits: Call functions with original sample input values

`_test_math_view` currently calls the operator with different values
than those specified in the `SampleInput`. This is undesirable as it
could break mathematical properties required by the operator. Instead,
this calls `math_op_view(math_op_physical(sample.input))` to get a
view that represents the same value as the original input.

`test_neg_view` already did this by returning `torch._neg_view(-x)`
from `math_op_view` but this moves the handling into `_test_math_view`
to make it apply to all view op tests.

[ghstack-poisoned]

peterbell10 committed:
Update on "TestMathBits: Call functions with original sample input values"


`_test_math_view` currently calls the operator with different values
than those specified in the `SampleInput`. This is undesirable as it
could break mathematical properties required by the operator. Instead,
this calls `math_op_view(math_op_physical(sample.input))` to get a
view that represents the same value as the original input.

`test_neg_view` already did this by returning `torch._neg_view(-x)`
from `math_op_view` but this moves the handling into `_test_math_view`
to make it apply to all view op tests.

[ghstack-poisoned]

anjali411 commented on test/test_ops.py:
```diff
@@ -1109,8 +1096,8 @@ def clone_and_perform_view(input, **kwargs):
-                        expected_forward.backward(math_op_physical(grad))
-                        forward_with_mathview.backward(math_op_view(grad))
+                        expected_forward.backward(grad)
+                        forward_with_mathview.backward(math_op_view(math_op_physical(grad)))
```
hmm we still want to test gradient computation when `grad_out` has a math bit set

peterbell10 replied:
This does have the math bit set. `math_op_physical` eagerly computes `t0 = f(grad)` and then `math_op_view` creates a view `t1 = f_view(t0)` and since the math ops are self  inverses `t1` represents the same value as the original `grad` but as a view.

anjali411 reviewed (commented):
Left one comment but looks good otherwise

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33064327).

anjali411 commented:
@peterbell10 phabricator is suggesting that there might be some merge conflicts due to recently landed PRs. can you please rebase this stack?

peterbell10 committed:
Update on "TestMathBits: Call functions with original sample input values"


`_test_math_view` currently calls the operator with different values
than those specified in the `SampleInput`. This is undesirable as it
could break mathematical properties required by the operator. Instead,
this calls `math_op_view(math_op_physical(sample.input))` to get a
view that represents the same value as the original input.

`test_neg_view` already did this by returning `torch._neg_view(-x)`
from `math_op_view` but this moves the handling into `_test_math_view`
to make it apply to all view op tests.

Differential Revision: [D33064327](https://our.internmc.facebook.com/intern/diff/D33064327)

[ghstack-poisoned]

peterbell10 commented:
@anjali411 rebased on master
